fsspec/implementations/tar.py

```diff
@@ -88,22 +88,25 @@ def _index(self):
         out = {}
         for ti in self.tar:
             info = ti.get_info()
-            info["type"] = typemap[info["type"]]
+            info["type"] = typemap.get(info["type"], "file")
             name = ti.get_info()["name"].rstrip("/")
             out[name] = (info, ti.offset_data)
 
         self.index = out
         # TODO: save index to self.index_store here, if set
 
     def _get_dirs(self):
-
         if self.dir_cache is not None:
             return
 
-        self.dir_cache = {}
+        # This enables ls to get directories as children as well as files
+        self.dir_cache = {
+            dirname + "/": {"name": dirname + "/", "size": 0, "type": "directory"}
+            for dirname in self._all_dirnames(self.tar.getnames())
+        }
         for member in self.tar.getmembers():
             info = member.get_info()
-            info["type"] = typemap[info["type"]]
+            info["type"] = typemap.get(info["type"], "file")
```

Comment:
Are there other "type"s that we should be inferring?

Reply:
There are a *few* inconsistencies between the various implementations, depending on when they were made. We still lack a comprehensive test system that guarantees that the API of each acts the same across all the implementations.

Reply:
👍 , got it. Odd!